src/azure-cli-core/azure/cli/core/parser.py

```diff
@@ -50,6 +61,185 @@ def _get_completions(self, comp_words, cword_prefix, cword_prequote, last_wordbr
                                                                                        last_wordbreak_pos)
 
 
+class CommandRecommender():
+    """Recommend a command for user when user's command fails.
+    It combines alladin recommendations and examples in help files."""
+
+    def __init__(self, command, parameters, extension, cli_ctx):
+        """
+        :param command: The command name in user's input, which should be corrected if misspelled.
+        :type command: str
+        :param parameters: The parameter arguments in users input.
+        :type parameters: list
+        :param extension: The extension name in user's input if the command comes from an extension.
+        :type extension: str
+        :param cli_ctx: CLI context when parser fails.
+        :type cli_ctx: knack.cli.CLI
+        """
+        self.command = command.strip()
+        self.extension = extension
+        self.cli_ctx = cli_ctx
+
+        self.parameters = self._normalize_parameters(parameters)
+        self.help_examples = []
+        self.aladdin_recommendations = []
+
+    def set_help_examples(self, examples):
+        """Set recommendations from help files"""
+
+        self.help_examples.extend(examples)
+
+    def _set_aladdin_recommendations(self):
+        """Set recommendations from aladdin service.
+        Call the aladdin service API, parse the response and set the recommendations.
+        """
+
+        import json
+        import requests
+        from requests import RequestException
+        from http import HTTPStatus
+        from azure.cli.core import __version__ as version
+
+        api_url = 'https://app.aladdin.microsoft.com/api/v1.0/suggestions'
```

Comment:
Correct, the service will not exist in the air-gapped clouds. We are working on a way to put the model into air-gapped clouds so clients could download it and query it manually. Until that works, we should have a way to disable pinging the service if there are multiple failures. It could be a timed disabling (so users who loose internet for a time don't have it disabled forever).